src/uct/ib/rc/verbs/rc_verbs_ep.c

```diff
@@ -15,11 +15,82 @@
 #include <ucs/arch/bitops.h>
 #include <uct/ib/base/ib_log.h>
 
+#define UCT_RC_VERBS_CHECK_FENCE(_iface, _ep) \
+    do { \
+        ucs_status_t _stat; \
+        _stat = uct_rc_verbs_ep_is_fence(_iface, _ep); \
+        if (_stat != UCS_OK) { \
+            return _stat; \
+        } \
+    } while (0)
+
+#define UCT_RC_VERBS_CHECK_RES(_iface, _ep) \
+    UCT_RC_CHECK_RES(&(_iface)->super, &(_ep)->super) \
+    UCT_RC_VERBS_CHECK_FENCE(_iface, _ep)
+
+#define UCT_RC_VERBS_CHECK_RES_FC(_iface, _ep) \
+    UCT_RC_CHECK_RES(&(_iface)->super, &(_ep)->super) \
+    UCT_RC_CHECK_FC(&iface->super, &ep->super, id) \
+    UCT_RC_VERBS_CHECK_FENCE(_iface, _ep)
+
 void uct_rc_verbs_txcnt_init(uct_rc_verbs_txcnt_t *txcnt)
 {
     txcnt->pi = txcnt->ci = 0;
 }
 
+static ucs_status_t
+uct_rc_verbs_ep_is_fence(uct_rc_verbs_iface_t *iface, uct_rc_verbs_ep_t *ep)
+{
+    ucs_status_t status;
+
+    if (iface->super.config.fence_mode != UCT_RC_FENCE_MODE_STRONG) {
+        return UCS_OK;
+    }
+
+    if (!uct_rc_ep_is_fence(&iface->super, &ep->fi)) {
+        return UCS_OK;
+    }
+    
+    if (iface->super.tx.fi.fence_beat - ep->fi.fence_beat == UCT_RC_FENCE_INPROGRESS) {
+        /* flush op is posted */
+        if (uct_rc_txqp_available(&ep->super.txqp) != iface->config.tx_max_wr) {
+            /* not all outstanding ops are completed */
+            return UCS_ERR_NO_RESOURCE;
+        } else {
+            /* fence is completed */
+            uct_rc_ep_fm_reset(&iface->super, &ep->fi);
+            return UCS_OK;
+        }
+    }
+
+    /* iface is scheduled and no flush request is posted.
```

Comment:
what does `iface is scheduled` mean?

Reply:
copy/paste issue: fence is scheduled
fixed